base/dsfs/commit.go

```diff
@@ -95,13 +100,89 @@ func commitFileAddFunc(privKey crypto.PrivKey) addWriteFileFunc {
 	}
 }
 
-// generateCommitTileAndMessage creates the commit and title, message
-func generateCommitTitleAndMessage(ctx context.Context, fs qfs.Filesystem, privKey crypto.PrivKey, ds, prev *dataset.Dataset, bodyAct BodyAction, fileHint string, forceIfNoChanges bool) error {
-	log.Debugf("generateCommitTitleAndMessage bodyAct=%s", bodyAct)
+// confirmByteChangesExist returns an early error if no components paths
+// differ from the previous flag & we're not forcing a commit.
+// if we are forcing a commit, set commit title and message values, which
+// triggers a fast-path in ensureCommitTitleAndMessage
+//
+// keep in mind: it is possible for byte-level changes to exist, but not cause
+// any alterations to dataset values, (for example: removing non-sensitive
+// whitespace)
+func confirmByteChangesExist(ds, prev *dataset.Dataset, added map[string]string, bodyPathName string, force bool) error {
+	if force {
+		log.Debugf("forcing changes. skipping uniqueness checks")
+		// fast path: forced changes ignore all comparison
+		if ds.Commit.Title == "" {
+			ds.Commit.Title = "forced update"
+		}
+		if ds.Commit.Message == "" {
+			ds.Commit.Message = "forced update"
+		}
+		return nil
+	}
+
+	if prev == nil {
+		return nil
+	}
+
+	// Viz, Readme and Transform components are inlined in the dataset, so they
+	// don't have path values before the commit component is finalized.
+	// use field equality checks instead of path comparison
+	if !ds.Viz.ShallowCompare(prev.Viz) {
+		log.Debugf("byte changes exist. viz components are inequal")
+		return nil
+	}
+	if !ds.Readme.ShallowCompare(prev.Readme) {
+		log.Debugf("byte changes exist. readme components are inequal")
+		return nil
+	}
+	if !ds.Transform.ShallowCompare(prev.Transform) {
+		log.Debugf("byte changes exist. transform components are inequal")
+		return nil
+	}
```

Comment:
What about structure (and should add future stats). I know they are qri calculated and not user supplied. Or maybe just hold that off for the change reports and use that for a medium level check before diving into diffing.

Reply:
structure and stats both have path values at this point, and are covered below in the `PathMap` comparisons